src/lib/core/backtrace.h

```diff
@@ -1,64 +1,106 @@
- * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
- * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
- * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
- * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
- * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
- * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
- * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
- * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
- * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
- * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
- * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
- * SUCH DAMAGE.
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
  */
+
+#pragma once
+
 #include "trivia/config.h"
-#include <stddef.h>
+#include "trivia/util.h"
 
-#if defined(__cplusplus)
-extern "C" {
-#endif /* defined(__cplusplus) */
+#include "libunwind.h"
 
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
 #ifdef ENABLE_BACKTRACE
-#include <coro.h>
+/*
+ * Format for printing Core frames.
+ */
+extern const char *const core_frame_str_fmt;
 
-char *
-backtrace(char *start, size_t size);
+enum {
+	/* Maximal number of Core frames collected. */
+	CORE_BACKTRACE_FRAME_COUNT_MAX = 64,
+};
 
-void print_backtrace(void);
+/*
+ * Core frame information sufficient for further resolving of function name
+ * and offset.
+ */
+struct core_frame {
+	/* Value of IP register. */
+	void *ip;
+};
 
-typedef int (backtrace_cb)(int frameno, void *frameret,
-                           const char *func, size_t offset, void *cb_ctx);
+/*
+ * Collection of Core frames.
+ */
+struct core_backtrace {
+	/* Number of Core frames collected. */
+	int frame_count;
+	/* Array of Core frames to be resolved. */
+	struct core_frame frames[CORE_BACKTRACE_FRAME_COUNT_MAX];
+};
 
+struct fiber;
 
-void
-backtrace_foreach(backtrace_cb cb, coro_context *coro_ctx, void *cb_ctx);
+/*
+ * Callback for processing Core frames.
+ *
+ * Returns status:
+ * 0 - continue processing.
+ * -1 - stop processing
+ */
+typedef int
+(*core_frame_cb)(const struct core_frame *, int frame_no, const char *proc_name,
+		 size_t offset, void *ctx);
 
+/*
+ * Collect call stack of `fiber` (only C/C++ frames) to `bt`.
+ *
+ * Nota bene: requires its own stack frame — hence, NOINLINE.
+ */
+NOINLINE void
+core_backtrace_collect_frames(struct core_backtrace *bt,
+			      const struct fiber *fiber,
+			      int skip_frames /* = 2 by default. */);
+
+/*
+ * Process Core frame.
+ *
+ * An additional buffer is needed for demangling C++ names, see `output_buf`
+ * and `length` parameters of `abi::__cxa_demangle`: https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a01696.html.
+ *
+ * The caller is responsible for freeing the pointer stored in `demangle_buf`.
+ * `demangle_buf` and `demangle_buf_len` can be reused for multiple calls.
```

Comment:
Do we really need to pass this buffer around? Can't we make it a static per-thread variable defined in `backtrace.cc` to simplify the function API?

Reply:
Fixed:
* see response above about `abi::__cxa_demangle`;
* encapsulated `demangle_buf` and `demangle_buf_len` into `thread_local` variables in `cxx_abi.cc`.